File: python/cote/LTCD/구간_병합.py
from typing import List
import random

TEST_CASES = 10
N = 10000
MAX_NUMBER = 10000
MAX_INTERVAL = 100

def initialize() -> List[List[int]] :
	ret = []
	n = random.randrange(2, N)

	for _ in range(n) :
		hi = random.randrange(MAX_INTERVAL, MAX_NUMBER)
		lo = hi - random.randrange(1, MAX_INTERVAL)
		ret.append([lo, hi])

	ret.sort()
	return ret

def solve(intervals: List[List[int]]) -> List[List[int]] :
	merged = []

	for cur in sorted(intervals, key = lambda x : x[0]) :
		if merged and cur[0] <= merged[-1][1] :
			merged[-1][1] = max(merged[-1][1], cur[1])
		else :
			merged.append(cur)

	return merged

if __name__ == "__main__" :
	random.seed(43)

	for _ in range(TEST_CASES) :
		intervals = initialize()
		print(solve(intervals))

File: javascript/cote/LTCD/구간_병합.mjs
const TEST_CASES = 1
const N = 10000
const MAX_NUMBER = N
const MAX_INTERVAL = 100

function initialize() {
	const n = Math.floor(Math.random() * (N - 1)) + 1
	const ret = []

	for (let i = 0; i < n; i++) {
		const begin = Math.floor(Math.random() * (MAX_NUMBER - MAX_INTERVAL)) 
		const end = begin + Math.floor(Math.random() * MAX_INTERVAL)
		ret.push([begin, end])
	}

	return ret
}

function solve(intervals) {
	const merged = []

	intervals.sort((a, b) => a[0] - b[0])
	console.log(intervals)

	for (const [begin, end] of intervals) {
		const last = merged ? merged[merged.length - 1] : undefined
		if (last && begin <= last[1] ) {
			last[1] = Math.max(last[1], end)
		} else {
			merged.push([begin, end])
		}

	}

	return merged
}

function main () {
	for(let i = 0; i < TEST_CASES; i++) {
		const intervals = initialize()
		console.log(intervals)
		console.log(solve(intervals))
	}
}

main()